src/tracer.h

```diff
@@ -174,26 +215,84 @@ class TraceManager {
         std::chrono::duration_cast<std::chrono::nanoseconds>(
             std::chrono::steady_clock::now().time_since_epoch());
 
-    opentelemetry::nostd::shared_ptr<otel_trace_api::Span> trace_span_{nullptr};
-
     std::unique_ptr<otel_trace_sdk::SpanExporter> exporter_;
 
     std::unique_ptr<otel_trace_sdk::SpanProcessor> processor_;
 
     std::shared_ptr<otel_trace_api::TracerProvider> provider_;
 
-    // Initializes Opentelemetry exporter, processor and provider
-    void InitTracer(const TraceConfigMap& config_map);
-
-    // Starts a span with the provided timestamp
-    void InitSpan(const opentelemetry::common::SystemTimestamp& timestamp_ns);
-
-    // Ends the started span and cleans up tracer
-    void EndSpan();
+    // OTel context to store spans, created in the current trace
+    opentelemetry::context::Context otel_context_;
+
+    /// Starts a span with the provided timestamp.
+    ///
+    /// \param display_name Span's name, which will be shown in the trace.
+    /// \param raw_timestamp_ns Steady timestamp, which is used to calculate
+    /// OpenTelemetry SystemTimestamp to display span on a timeline, and
+    /// OpenTelemetry SteadyTimestamp to calculate the duration on the span
+    /// with better precision.
+    /// \param is_root_span If true, a root span will be started,
+    /// i.e. with no parent span specified. If false, a new child span will
+    /// be started.
+    /// \param parent_span_key A span key, to find a parent span in the
+    /// OpenTelemetry context.
+    /// \return A shared pointer to a newly created OpenTelemetry span.
+    opentelemetry::nostd::shared_ptr<otel_trace_api::Span> StartSpan(
+        std::string display_name, const uint64_t& raw_timestamp_ns,
+        bool is_root_span, std::string parent_span_key = "");
+
+    /// Ends the provided span.
+    ///
+    /// \param span_key Span's key to retrieve the corresponding span from the
+    /// OpenTelemetry context.
+    void EndSpan(std::string span_key);
+
+    /// Ends the provided span at specified steady timestamp.
+    ///
+    /// \param span_key Span's key to retrieve the corresponding span from the
+    /// OpenTelemetry context.
+    /// \param raw_timestamp_ns Steady timestamp to use as
+    /// `EndSpanOptions::end_steady_time`.
+    void EndSpan(std::string span_key, const uint64_t& raw_timestamp_ns);
```

Comment:
I am assuming no one would want to EndSpan with `raw_timestamp_ns=0`? 
We might be able to use same function signature -  if raw_timestamp_ns is 0, then we can record the current time.

Reply:
Ignore this. It looks good as is.